tag-system/src/main/java/org/togetherjava/tjbot/tag/TagSystemInstanceHolder.java

```diff
@@ -0,0 +1,33 @@
+package org.togetherjava.tjbot.tag;
+
+import org.togetherjava.tjbot.db.Database;
+
+public class TagSystemInstanceHolder {
+    private TagSystemInstanceHolder() {}
+
+    private static volatile TagSystem instance;
+
+    public static TagSystem getOrCreate(Database database) {
+        if (instance == null) {
+            setInstance(new TagSystem(database));
```

Comment:
is lazy init worth the trouble of making this class safe? why not just have the instance always?

Reply:
In particular, i dislike that the two methods below can only give runtime safety and can not prevent wrong calls at compile-time.

without lazy init, all of those trouble is gone.

Reply:
> is lazy init worth the trouble of making this class safe? why not just have the instance always?

Because otherwise there will be no way to retrieve the database instance at the beginning.

Reply:
Makes sense, thanks for explaining.

Reply:
Any reason why you need a singleton though - why not just a plain regular instance?

Reply:
> Any reason why you need a singleton though - why not just a plain regular instance?

Everyone would need to pass the instance around and store it somewhere, possibly creating circular dependencies.

Reply:
Why would there be cyclic references? You create the system in `main` and pass it around where needed?

Reply:
removed the singleton